sdk/src/main/java/io/opentelemetry/sdk/metrics/AbstractBoundInstrument.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.metrics;
+
+import io.opentelemetry.metrics.LabelSet;
+import java.util.concurrent.atomic.AtomicLong;
+
+abstract class AbstractBoundInstrument {
+  private final LabelSet labels;
+  // Atomically counts the number of references (usages) while also keeping a state of
+  // mapped/unmapped into a registry map.
+  private final AtomicLong refCountMapped;
+  private final Aggregator<?> aggregator;
+
+  AbstractBoundInstrument(LabelSet labels, Aggregator<?> aggregator) {
+    this.labels = labels;
+    this.aggregator = aggregator;
+    refCountMapped = new AtomicLong(0);
+  }
+
+  /**
+   * Returns {@code true} if the entry is still mapped and increases the reference usages, if
+   * unmapped returns {@code false}.
+   *
+   * @return {@code true} if successful.
+   */
+  boolean ref() {
+    return (refCountMapped.addAndGet(2L) & 1L) == 0;
```

Comment:
2?

Reply:
I kinda see what's going on here, but cleverness like this needs some very clear documentation in the code.

Reply:
will do, sorry this is still work in progress as the PR description says.

Reply:
yep! no worries, just trying to capture my thoughts so we don't lose them.  :)

Reply:
done.